grl/mi.py

```diff
@@ -0,0 +1,207 @@
+                                   pi_lr=pi_lr,
+                                   mi_lr=mi_lr,
+                                   mi_iterations=mi_iterations,
+                                   pi_per_step=pi_steps,
+                                   mi_per_step=mi_steps)
+
+    info['initial_policy'] = initial_policy
+    # we get lambda discrepancies here
+    # initial policy lambda-discrepancy
+    info['initial_policy_stats'] = lambda_discrep_measures(amdp, initial_policy)
+    info['initial_improvement_stats'] = lambda_discrep_measures(amdp,
+                                                                info['initial_improvement_policy'])
+    greedy_initial_improvement_policy = greedify(info['initial_improvement_policy'])
+    info['greedy_initial_improvement_stats'] = lambda_discrep_measures(
+        amdp, greedy_initial_improvement_policy)
+
+    # Initial memory amdp w/ initial improvement policy discrep
+    if 'initial_mem_params' in info and info['initial_mem_params'] is not None:
+        init_mem_amdp = memory_cross_product(amdp, info['initial_mem_params'])
+        info['initial_mem_stats'] = lambda_discrep_measures(
+            init_mem_amdp, info['initial_expanded_improvement_policy'])
+
+    # Final memory w/ final policy discrep
+    final_mem_amdp = memory_cross_product(amdp, agent.mem_params)
+    info['final_mem_stats'] = lambda_discrep_measures(final_mem_amdp, agent.policy)
+    greedy_final_policy = greedify(agent.policy)
+    info['greedy_final_mem_stats'] = lambda_discrep_measures(final_mem_amdp, greedy_final_policy)
+
+    return info, agent
+
+def memory_iteration(
+    agent: AnalyticalAgent,
+    init_amdp: AbstractMDP,
+    pi_lr: float = 1.,
+    mi_lr: float = 1,
+    pi_per_step: int = 50000,
+    mi_per_step: int = 50000,
+    mi_iterations: int = 1,
+    init_pi_improvement: bool = True,
+    log_every: int = 1000,
+):
+    info = {'policy_improvement_outputs': [], 'mem_loss': []}
+    td_v_vals, td_q_vals = td_pe(agent.policy, init_amdp.T, init_amdp.R, init_amdp.phi,
+                                 init_amdp.p0, init_amdp.gamma)
+    info['initial_values'] = {'v': td_v_vals, 'q': td_q_vals}
+
+    if init_pi_improvement:
+        # initial policy improvement
+        print("Initial policy improvement step")
+        initial_outputs = pi_improvement(agent,
+                                         init_amdp,
+                                         lr=pi_lr,
+                                         iterations=pi_per_step,
+                                         log_every=log_every)
+        info['policy_improvement_outputs'].append(initial_outputs)
+
+    info['initial_improvement_policy'] = agent.policy.copy()
+    info['initial_mem_params'] = agent.mem_params
+    print(f"Starting (unexpanded) policy: \n{agent.policy}")
+
+    # we have to set our policy over our memory MDP now
+    # we do so with a random policy given new memory bit
+    agent.new_pi_over_mem()
```

Comment:
The comment might be inaccurate, since there's a `"random"` mode and a `"copy"` mode, and `"copy"` is the default